File: C#/CookieStack.cs
using System;
using System.Collections;

/*
    CookieDataStructure
    Copyright (C) 2026 Pe4enbka008 (Helen Ivanova) 

    This code is free software: you can redistribute it and/or modify 
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or any later version. 

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
*/



/*
CookieDataStructure: CookieStack contains
    IsEmpty - public function
    Push - public function
    PopValue - public function
    GetTop - public function
    Copy - public function
    Clear - public function
    Reverse - public function
    RotateLeft - public function
    RotateRight - public function
    GetEnumerator - public function (and IEnumerable.GetEnumerator implementation)
    ToString - public function (2 overloads: no-arg, with split string)
*/


namespace smth
{
    /// <summary>
    /// This CookieDataStructure requires CookieNode.cs file!
    /// Stack created and better-ed by Cookie :]
    /// </summary>
    /// <typeparam name="StackType">Type of the stack</typeparam>
    public class CookieStack<StackType> : IEnumerable<StackType>
    {
        private CookieNode<StackType>? head_node;

        private static string className = "CookieStack";

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack() { this.head_node = null; }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(StackType value) { this.head_node = new CookieNode<StackType>(value); }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(CookieNode<StackType>? nodes) { this.head_node = nodes; }

        /// <summary>
        /// Class constructor
        /// </summary>
        public CookieStack(StackType[] list) { this.head_node = CookieNodeWorker.ArrayToNodes(list); }


        /// <summary>
        /// Returns if the stack is empty
        /// </summary>
        /// <returns>true is the stack is empty</returns>
        public bool IsEmpty() { return this.head_node == null; }

        /// <summary>
        /// Returns stack's length 
        /// </summary>
        public int Length { get { return CookieNodeWorker.RecursionCount<StackType>(this.head_node); } }


        // Getters-Setters
        /// <summary>
        /// FILO (FIRST IN ; LAST OUT) - puts the value in front
        /// </summary>
        /// <param name="value">Value to save</param>
        public void Push(StackType value) 
        { this.head_node = new CookieNode<StackType>(value, this.head_node); }


        /// <summary>
        /// LIFO (LAST IN ; FIRST OUT) - gets the front value ; could be null
        /// </summary>
        /// <returns>The top (last added) value</returns>
        public StackType PopValue()
        {
            if (this.head_node == null)
                throw new CookieEmptyStructureException(className);

            CookieNode<StackType> node = this.head_node;
            this.head_node = node.GetNext();
            node.Next = null;
            return node.Value;
        } // Pop

        /// <summary>
        /// Easier way of Pop function :]
        /// </summary>
        public StackType Pop { get { return this.PopValue(); } }


        /// <summary>
        /// Gets the top value, if the stack is empty, returns default of the type
        /// </summary>
        /// <returns>value of the top</returns>
        public StackType GetTop()
        { return this.head_node != null ? this.head_node.Value : throw new CookieEmptyStructureException(className); } 


        /// <summary>
        /// Creates a copy of the object
        /// </summary>
        /// <returns>copy of the node list</returns>
        public CookieStack<StackType> Copy()
        {
            if (this.head_node == null)
                throw new CookieEmptyStructureException(className);

            CookieNode<StackType> return_value = new CookieNode<StackType>(this.head_node.Value);
            CookieNode<StackType> current = return_value;
            CookieNode<StackType>? nodes = this.head_node.Next;

            while (nodes != null)
            {
                current.Next = new(nodes.Value);
                current = current.Next;
                nodes = nodes.Next;
            } // while

            return new(return_value);
        } // Copy


        /// <summary>
        /// Clears the Stack
        /// </summary>
        public void Clear()
        { this.head_node = null; }


        /// <summary>
        /// Creates a copy of the object in type of the CookieStack
        /// </summary>
        /// <returns>copy of the node list</returns>
        public CookieStack<StackType> Reverse()
        {
            if (this.Length <= 1) return this.Copy();

            CookieStack<StackType> rev = new CookieStack<StackType>();
            CookieNode<StackType>? current = this.head_node;

            while (current != null)
            {
                rev.Push(current.Value);
                current = current.Next;
            } // while
            return rev;
        } // Reverse



        /// <summary>
        /// Moves front to back once
        /// </summary>
        public void RotateLeft()
        { if (!this.IsEmpty()) this.Push(this.Pop); }

        /// <summary>
        /// Moves back to front once
        /// </summary>
        public void RotateRight()
        {
            if (this.IsEmpty() || this.head_node.Next == null)  // there is one value in
                return;

            CookieNode<StackType> current = this.head_node;
            while (current.Next.Next != null)
                current = current.Next;
            CookieNode<StackType> last = current.Next;
            current.SetNext(null);

            last.SetNext(this.head_node);
            this.head_node = last;
        } // RotateRight



        // override

        // IEnumerable
        /// <summary>
        /// Returns an enumerator that walks the stack from top to bottom (enables foreach)
        /// </summary>
        /// <returns>An enumerator over the stack's values, top first</returns>
        public IEnumerator<StackType> GetEnumerator()
        { if (this.head_node != null) return this.head_node.GetEnumerator(); return Enumerable.Empty<StackType>().GetEnumerator(); }
        /// <summary>
        /// Non-generic IEnumerable.GetEnumerator implementation, forwards to the generic version
        /// </summary>
        /// <returns>An enumerator over the stack's values, top first</returns>
        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();


        // object 

        /// <summary>
        /// override for ToString to - ['value', 'value', 'value', ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public override string ToString()
        {
            return ToString(", ");
        } // override ToString

        /// <summary>
        /// override for ToString to - ['value'{split}'value'{split}'value'{split} ...]
        /// </summary>
        /// <returns>string of the class</returns>
        public string ToString(string split)
        {
            if (this.head_node == null)
                return "[]";

            string str = "[";
            CookieNode<StackType>? node = this.head_node;

            while (node != null)
            {
                str += node.ToString();
                if (node.GetNext() != null)
                    str += split;
                node = node.GetNext();
            } // while

            return str + "]";
        } // override ToString

    } // class CookieStack



} // namespace smth

File: Python/CookieStack.py
from CookieNode import CookieNode
import CookieErrors

"""
    CookieDataStructure
    Copyright (C) 2026 Pe4enbka008 (Helen Ivanova)

    This code is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

    See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program. If not, see <https://www.gnu.org/licenses/>.
"""


class CookieStack:
    def __init__(self, value=None):
        """Constructor for CookieStack
        :param value: start nodes
        :type value: CookieNode | CookieStack | Any
        :return: nothing
        :rtype: None"""
        self.class_name = 'CookieStack'
        self.nodes = None

        if isinstance(value, CookieNode):
            self.nodes = value
        elif isinstance(value, CookieStack):
            for node in value.nodes:
                self.push(node.get_value())
        elif value is not None:
            self.nodes = CookieNode(value)

    def __str__(self):
        """Returns string value of the nodes
        :return: string value of the nodes
        :rtype: str"""
        return self.to_str(', ')

    def to_str(self, spliter):
        """Returns string value of the nodes
        :param spliter: spliter to put in between of the values
        :type spliter: str
        :return: string value of the nodes
        :rtype: str"""
        if not self.nodes:
            return '[]'

        return_value = '['
        walk_value = self.nodes
        while walk_value:
            return_value += str(walk_value)
            walk_value = walk_value.get_next()
            if walk_value:
                return_value += spliter

        return return_value + ']'

    def is_empty(self):
        """Returns true if the stack is empty
        :return: true if the stack is empty
        :rtype: bool"""
        return not self.nodes

    # getters-setters:

    def push(self, value):
        """FILO (FIRST IN ; LAST OUT) - puts the value in front
        :param value: value to push
        :type value: CookieNode | CookieStack | Any
        :return: nothing
        :rtype: None"""
        if isinstance(value, CookieNode):
            value.next_node = self.nodes
            self.nodes = value
        elif isinstance(value, CookieStack):
            for node in value.nodes:
                self.push(node.get_value())
        else:
            self.nodes = CookieNode(value, self.nodes)

    def pop(self):
        """LIFO (LAST IN ; FIRST OUT) - gets the front value ; could be null
        :return: value to pop
        :rtype: Any"""
        if self.is_empty():
            raise CookieErrors.CookieEmptyStructureException(self.class_name)

        node = self.nodes
        self.nodes = self.nodes.next_node
        node.next_node = None
        return node.value

    def get_top(self):
        """Returns the top value of the stack
        :return: the top value of the stack
        :rtype: Any"""
        if self.nodes:
            return self.nodes.value
        raise CookieErrors.CookieEmptyStructureException(self.class_name)

    def copy(self):
        """Returns a copy of the stack
        :return: the copy of the stack
        :rtype: CookieStack"""
        if not self.nodes:
            raise CookieErrors.CookieEmptyStructureException(self.class_name)

        return_value = CookieNode(self.nodes.value)
        current = return_value
        walk_value = self.nodes.next_node

        while walk_value:
            current.set_next(CookieNode(walk_value.value))
            current = current.get_next()
            walk_value = walk_value.get_next()

        return CookieStack(return_value)

    def clear(self):
        """Clears the stack
        :return: nothing
        :rtype: None"""
        self.nodes = None

    # length:
    def __len__(self):
        """Returns length of the stack
        :return: length
        :rtype: int"""
        if self.is_empty():
            return 0
        return sum(1 for _ in self.nodes)

    # reverse:
    def __reversed__(self):
        """reverses a stack
        :return: reversed stack
        :rtype: CookieStack"""
        if len(self) <= 1:
            return self.copy()

        current = self.nodes
        rev = CookieStack()
        while current:
            rev.push(current.get_value())
            current = current.get_next()
        return rev

    # end

